snapcraft/services/lifecycle.py

```diff
@@ -15,17 +15,49 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """Snapcraft Lifecycle Service."""
 
-from typing import cast
+import copy
+import os
+import shutil
+from datetime import datetime
+from pathlib import Path
+from typing import Any, cast
 
-from craft_application import LifecycleService
+from craft_application import AppMetadata, LifecycleService, ServiceFactory
+from craft_application.models import BuildInfo
+from craft_cli import emit
 from overrides import overrides
 
-from snapcraft import models
+from snapcraft import models, utils
+from snapcraft.meta import manifest
 
 
 class Lifecycle(LifecycleService):
     """Snapcraft specialization of the Lifecycle Service."""
 
+    def __init__(  # noqa: PLR0913 (too many arguments)
+        self,
+        app: AppMetadata,
+        services: ServiceFactory,
+        *,
+        project: models.Project,
+        work_dir: Path | str,
+        cache_dir: Path | str,
+        build_plan: list[BuildInfo],
+        project_path: Path | None,
```

Comment:
Seems like `project_path` should always be a `Path`. When could it validly be `None`?

Reply:
Never. At this point it must always be valid.